### DISCUSSION TITLE
CloudFlare Worker is great BUT if on Pages it would be fantastic !

### DISCUSSION THREAD

**Discussion by FabriceACohen** [protomaps/PMTiles#224] [Ideas]
Hello,

I love your work. And just implemented it on Cloudflare. Now CloudFlare have opened Pages that are even less expensive than Workers and allow to put everything at the same place : function.js, style.json, glyths, and fonts. All that for 0$.

I tried to tweak it from the worker.js into _worker.js as adviced by Cloudflare (Refactor a Worker to a Pages Function : https://developers.cloudflare.com/pages/how-to/refactor-a-worker-to-pages-functions/) but got exeptions and the style.json from pages is not even provided anymore. 

I am sure it is close but still blocked.

What I have done so far :

- created a forlder on my desktop with _worker.js (copy of your [index.js](https://protomaps.github.io/PMTiles/index.js) and added my style.json file.
- deployed this folder in a new projet
- in the project "Settings"=>"Functions"=> R2 bucket bindings => Add Variable BUCKET to my_bucket
- in the project "Custom Domains" => Add custom domain

=> No luck. Got Exception Error 1101 Worker threw exception.
Looking in logs :  
  "diagnosticsChannelEvents": [],
  "exceptions": [
    {
      "name": "TypeError",
      "message": "Cannot read properties of undefined (reading 'get')",

Help would be great !

**Comment by bdon** [comment]
>  Now CloudFlare have opened Pages that are even less expensive than Workers

Can you link where you see details of this? According to https://developers.cloudflare.com/pages/platform/functions/pricing/ "Requests to your Functions are billed as Cloudflare Workers requests. "

Given that, I think it makes more sense to have tiles served from a Worker as usual, and you can use Pages for the other static parts of the application like frontend, fonts, etc. That way you can also have one tile Worker serving multiple applications instead of 1:1.

**Reply by FabriceACohen** [reply]
You are correct. That's the way I do it since this message.
